### PR TITLE
Re-introduce the spec compliance fix that removes the unspecced `room_id` field in the `/hierarchy` response.

### LINKED ISSUE #10946: Stabilize support for MSC2946: spaces summary
Once [MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946) lands we should (at some point) switch to a stable endpoint and make some updates. There's few bits to this that we need to do:

## Phase 1

* [x] Support to the stable endpoint.
* [x] ~~Require `limit` to be > 0.~~ (This has always been correct.)
* [x] Ensure the endpoint is ratelimited.

## Phase 2

* [x] Remove the fallback logic of `get_room_hierarchy` to request the `/spaces` endpoint over federation.
* [x] Remove the `/spaces` client and federation endpoints (and corresponding code in the handlers).
* [x] Update the API responses to remove `allowed_spaces` (this was replaced with `allowed_room_ids`).
* [x] Clean-up code around max-children, which is not used by the `/hierarchy` endpoint.
* [x] Remove `room_id` from the serialized `children_state`, it is redundant and removed from the API response.

## Phase 3

* [x] Remove unstable `/hierarchy` client and federation endpoints.
* [x] Remove the typo `join_rules` field from the response.

### PR DESCRIPTION
**Pull request by reivilibre**
Original: #13365

Reverts: #13501

Fixes #10946

Blocked:
- [ ] Element iOS: fixed and given enough time to propagate (https://github.com/matrix-org/matrix-ios-sdk/pull/1547)

### CODE REVIEW

**Comment by erikjohnston** [comment]
Both iOS and Android have confirmed that they're happy for this to be merged!

**Comment by clokep** [comment]
> Both iOS and Android have confirmed that they're happy for this to be merged!

See https://github.com/matrix-org/synapse/issues/10946#issuecomment-1227613519 and subsequent comments. I think that we're probably OK to upgrade now.

File: changelog.d/13506.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Re-introduce the spec compliance fix that removes the unspecced `room_id` field in the `/hierarchy` response.
```
Comment (clokep) [inline changelog.d/13506.bugfix:1]: I think we likely want to use the original changelog again:

```suggestion
Fix a bug introduced in Synapse v1.41.0 where the `/hierarchy` API returned non-standard information (a `room_id` field under each entry in `children_state`).
```

**Comment by clokep** [comment]
> Sanity check: we've confirmed that element clients are okay with this change. Have we also highlighted this removal with the community? I think we usually ask in the Client Developers' room?

This is how we found the issue in the first place and Element iOS was the only one that came up.

**Review by DMRobertson — APPROVED** [review]
Let's get this into 1.68.

**Pull request merged**